In [ ]:
import pandas as pd
df = pd.read_csv("MCSDatasetNEXTCONLab.csv")


EXPLORATORY DATA ANALYSIS

In [ ]:
print(df)

         ID   Latitude  Longitude  Day  Hour  Minute  Duration  RemainingTime  \
0         1  45.442142 -75.303369    1     4      13        40             40   
1         1  45.442154 -75.304366    1     4      23        40             30   
2         1  45.442104 -75.303963    1     4      33        40             20   
3         1  45.441868 -75.303577    1     4      43        40             10   
4         2  45.447727 -75.147722    2    15      49        30             30   
...     ...        ...        ...  ...   ...     ...       ...            ...   
14479  3999  45.445303 -75.165596    2     1      18        20             20   
14480  3999  45.445574 -75.165168    2     1      28        20             10   
14481  4000  45.436682 -75.152416    0    12      21        30             30   
14482  4000  45.436978 -75.153278    0    12      31        30             20   
14483  4000  45.436983 -75.153240    0    12      41        30             10   

       Resources  Coverage 

In [ ]:
df.head()

ID   Latitude  Longitude  Day  Hour  Minute  Duration  RemainingTime  \
0   1  45.442142 -75.303369    1     4      13        40             40   
1   1  45.442154 -75.304366    1     4      23        40             30   
2   1  45.442104 -75.303963    1     4      33        40             20   
3   1  45.441868 -75.303577    1     4      43        40             10   
4   2  45.447727 -75.147722    2    15      49        30             30   

   Resources  Coverage  OnPeakHours  GridNumber  Ligitimacy  
0          9        91            0      131380           1  
1          9        91            0      131380           1  
2          9        91            0      121996           1  
3          9        91            0      121996           1  
4          5        47            0      140784           1

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14484 entries, 0 to 14483
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ID             14484 non-null  int64  
 1   Latitude       14484 non-null  float64
 2   Longitude      14484 non-null  float64
 3   Day            14484 non-null  int64  
 4   Hour           14484 non-null  int64  
 5   Minute         14484 non-null  int64  
 6   Duration       14484 non-null  int64  
 7   RemainingTime  14484 non-null  int64  
 8   Resources      14484 non-null  int64  
 9   Coverage       14484 non-null  int64  
 10  OnPeakHours    14484 non-null  int64  
 11  GridNumber     14484 non-null  int64  
 12  Ligitimacy     14484 non-null  int64  
dtypes: float64(2), int64(11)
memory usage: 1.4 MB


In [ ]:
df.shape

(14484, 13)

This dataset contains 14484 rows and 13 coloumns
END of EDA



In [ ]:
df.isnull().sum()

ID               0
Latitude         0
Longitude        0
Day              0
Hour             0
Minute           0
Duration         0
RemainingTime    0
Resources        0
Coverage         0
OnPeakHours      0
GridNumber       0
Ligitimacy       0
dtype: int64

In [ ]:
df.duplicated().sum()

0

DATA VISUALIZATION

In [ ]:
x = df.iloc[:,:-1]
y = df.iloc[:,-1:]

In [ ]:
x

ID   Latitude  Longitude  Day  Hour  Minute  Duration  RemainingTime  \
0         1  45.442142 -75.303369    1     4      13        40             40   
1         1  45.442154 -75.304366    1     4      23        40             30   
2         1  45.442104 -75.303963    1     4      33        40             20   
3         1  45.441868 -75.303577    1     4      43        40             10   
4         2  45.447727 -75.147722    2    15      49        30             30   
...     ...        ...        ...  ...   ...     ...       ...            ...   
14479  3999  45.445303 -75.165596    2     1      18        20             20   
14480  3999  45.445574 -75.165168    2     1      28        20             10   
14481  4000  45.436682 -75.152416    0    12      21        30             30   
14482  4000  45.436978 -75.153278    0    12      31        30             20   
14483  4000  45.436983 -75.153240    0    12      41        30             10   

       Resources  Coverage  OnPeakHours  GridNumber  
0              9        91            0      131380  
1              9        91            0      131380  
2              9        91            0      121996  
3              9        91            0      121996  
4              5        47            0      140784  
...          ...       ...          ...         ...  
14479         10        80            0      131397  
14480         10        80            0      131397  
14481          4        63            0      122015  
14482          4        63            0      122015  
14483          4        63            0      122015  

[14484 rows x 12 columns]

In [ ]:
y

Ligitimacy
0               1
1               1
2               1
3               1
4               1
...           ...
14479           1
14480           1
14481           1
14482           1
14483           1

[14484 rows x 1 columns]

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.8)

In [ ]:
from imblearn.over_sampling import RandomOverSampler
ov = RandomOverSampler(sampling_strategy= 'minority')
x_train_resampled, y_train_resampled = ov.fit_resample(x_train, y_train)

In [ ]:
from sklearn.ensemble import RandomForestClassifier
RF_model = RandomForestClassifier(random_state=42)
RF_model = RF_model.fit(x_train_resampled, y_train_resampled)
y_pred_RF = RF_model.predict(x_test)

<ipython-input-13-01e097f1716a>:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  RF_model = RF_model.fit(x_train_resampled, y_train_resampled)


In [ ]:
from sklearn.metrics import classification_report, ConfusionMatrixDisplay, accuracy_score, confusion_matrix
from sklearn.ensemble import VotingClassifier
from imblearn.over_sampling import RandomOverSampler

import plotly.express as express

from sklearn.preprocessing import MinMaxScaler
from keras.models import Sequential, Model
from keras.layers import Reshape
from keras.layers import Dense
from keras.layers import LeakyReLU
from keras.layers import BatchNormalization
from keras.layers import Input, Flatten

GAN MinMax Scaler

In [ ]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
x_train_scaled = scaler.fit_transform(x_train_resampled)
x_test_scaled = scaler.transform(x_test)

NameError: name 'x_test' is not defined

DISCRIMINATOR FUNC.

In [ ]:
def define_discriminator():
  model = Sequential()
  model.add(Flatten(input_shape=x_train_scaled.shape[1:]))
  def add_discriminator_block(neurons, alpha = 0.3):
    model.add(Dense(neurons))
    model.add(LeakyReLU(alpha))

  add_discriminator_block(256)
  add_discriminator_block(512)

  model.add(Dense(1, activation='sigmoid'))


  data = Input(shape=x_train_scaled.shape[1:])
  output = model(data)

  return Model(data, output)

In [ ]:
noise_shape = (100,)

GENERATOR FUNC.

In [ ]:
import numpy as np
def define_generator():
  model = Sequential()

  def add_generator_block(neurons, alpha = 0.3):
    model.add(Dense(neurons))
    model.add(LeakyReLU(alpha))
    model.add(BatchNormalization())

    model.add(Dense(256, input_shape= noise_shape))
    model.add(LeakyReLU(alpha=.3))
    model.add(BatchNormalization())

    add_generator_block(512)
    add_generator_block(1024)

    model.add(Dense(np.prod(x_train_scaled.shape[1:]), activation='tanh'))

    model.add(Reshape(x_train_scaled.shape[1:]))


    noise = Input(shape= noise_shape)
    data = model(noise)

    return Model(noise, data)

In [ ]:
def train(epochs, batch, generator, discriminator, combined):
    half_batch = int(batch / 2)
    for epoch in range(epochs):


      indices = np.random.randint(0, x_train_scaled.shape[0], half_batch)
      data = x_train_scaled[indices]


    noise = np.random.normal(0, 1, (half_batch, 100))
    fake_data = generator.predict(noise)

    discriminator_loss_real = discriminator.train_on_batch(data, np.ones((half_batch, 1)))
    discriminator_loss_fake = discriminator.train_on_batch(fake_data, np.zeros((half_batch, 1)))
    discriminator_loss = .5 * np.add(discriminator_loss_real, discriminator_loss_fake)


    noise = np.random.normal(0,1, (batch, 100))
    valid_y = np.array([1]*batch)
    generator_loss = combined.train_on_batch(noise, valid_y)

    print("epoch: %d Discriminator loss: %f - Generator loss: %f" % (epoch+1,  discriminator_loss[0], generator_loss))

    discriminator = define_discriminator()
    discriminator.compile(loss="binary_crossentropy", optimizer= "adam", metrics=["accuracy"])

    generator = define_generator()
    generator.compile(loss="binary_crossentropy", optimizer= "adam")

    noise = Input(shape=(100,))
    fake_data = generator(noise)

    discriminator.trainable = False

    valid = discriminator(fake_data)

    combined = Model(noise, valid)
    combined.compile(loss="binary_crossentropy", optimizer= "adam")

    train(500, 32, generator, discriminator, combined)
